src/arextract.py

```diff
@@ -106,6 +109,209 @@ def boxcar(slf, det, specobjs, sciframe, varframe, skyframe, crmask, scitrace):
+                mx = np.max(slit_val[gdp])
+                xval = np.linspace(mn,mx,1000)
+                model = arutils.func_val(gfit, xval, fdict['func'])
+                import matplotlib.pyplot as plt
+                plt.clf()
+                ax = plt.gca()
+                ax.scatter(slit_val[gdp], flux_val[gdp], marker='.', s=0.7, edgecolor='none', facecolor='black')
+                ax.plot(xval, model, 'b')
+                # Gaussian too?
+                if False:
+                    fdictg = dict(func='gaussian', deg=3)
+                    maskg, gfitg = arutils.robust_polyfit(slit_val, flux_val, fdict['deg'], function=fdictg['func'], weights=weight_val, maxone=False)
+                    modelg = arutils.func_val(gfitg, xval, fdictg['func'])
+                    ax.plot(xval, modelg, 'r')
+                plt.show()
+                debugger.set_trace()
+        elif len(gdrow) > 10:  #
+            msgs.warn("Low extracted flux for obj={:s}.  Not ready for Optimal".format(specobjs[o].idx))
+            scitrace['opt_profile'].append({})
+            continue
+    # QA
+    arqa.obj_profile_qa(slf, specobjs, scitrace)
+    return scitrace['opt_profile']
+
+
+def optimal_extract(slf, det, specobjs, sciframe, varframe, skyframe, crmask, scitrace,
+                 pickle_file=None, profiles=None):
+    """ Preform optimal extraction
+    Standard Horne approach
+    Parameters
+    ----------
+    slf
+    det
+    specobjs
+    sciframe
+    varframe
+    crmask
+    scitrace
+    COUNT_LIM
+    pickle_file
+    Returns
+    -------
+    """
+    '''
+    import pickle
+    if pickle_file is not None:
+        f = open(pickle_file,'r')
+        args = pickle.load(f)
+        f.close()
+        det, specobjs, sciframe, varframe, skyframe, crmask, scitrace, tilts = args
+        slf = None
+        scitrace['opt_profile'] = profiles
+    else:
+        tilts = None
+    '''
+    # Setup
+    model_var = np.abs(skyframe + sciframe - np.sqrt(2)*slf._msrn[det-1]) + \
+                 slf._msrn[det-1]**2  # sqrt 2 term deals with negative flux/sky
+    model_ivar = 1./model_var
+    msgs.work("Consider making a model of the object for model_ivar")
+    cr_mask = 1.0-crmask
+    # Loop
+    nobj = scitrace['traces'].shape[1]
+    for o in range(nobj):
+        # Fit dict
+        fit_dict = scitrace['opt_profile'][o]
+        if 'param' not in fit_dict.keys():
+            continue
+        # Slit image
+        slit_img = artrace.slit_image(slf, det, scitrace, o)#, tilts=tilts)
+        #msgs.warn("Turn off tilts")
+        # Object pixels
+        weight = scitrace['object'][:,:,o]
+        gdo = (weight > 0) & (model_ivar > 0)
+        # Profile image
+        prof_img = np.zeros_like(weight)
+        prof_img[gdo] = arutils.func_val(fit_dict['param'], slit_img[gdo],
+                                         fit_dict['func'])
+        # Normalize
+        norm_prof = np.sum(prof_img, axis=1)
```

Comment:
Overall, I agree.  Indeed, this is what I instituted with ESI extraction onto a 
fixed wavelength grid (although I took simple linear interpolation instead of
integrating over the pixel).

I don't recall why we took this step (backwards) in LowRedux.  Definitely
worth revisiting but I think we should implement as is now and then develop
a more sophisticated approach as desired.